### PR TITLE
Implement Saved Zoom Levels in Place List Pages

### LINKED ISSUE #3229: Geographic data: Some Maps display Incorrectly.
Greg as requested I have made a new WebTrees Test tree containing 17 individuals, and includes the problems with the maps I mentioned on the Forum.
https://webtrees.wilson-mciver.com/index.php?route=%2Ftree%2FWTtest

Search for Ann Greg or click here:-
https://webtrees.wilson-mciver.com/index.php?route=%2Ftree%2FWTtest%2Findividual%2FI6241%2FAnn-Gregg

If you click on her death place <Mirehouse, Keswick, Cumberland, England> the following map is displayed
 
image

**This is NOT the correct map**, I have accurately defined <Mirehouse>, under <Control panel / Geographic data / England / Cumberland / Keswick / Mirehouse>.
The following Map is the one that should be displayed here.
image

 If you click on her Burial place <St. Bega Church, Mirehouse, Keswick, Cumberland, England> the First Map above is correctly Displayed. 

=============================================
 

A similar situation exists for Roger Williamson
https://webtrees.wilson-mciver.com/index.php?route=%2Ftree%2FWTtest%2Findividual%2FI6232%2FRoger-Williamson
If you Click on <Baptism Torpenhow, Cumberland, England>, the following Map is displayed.
 
image

This isn’t correct as it doesn’t show <Torpenhow> the town which is nearby and is itself defined in the following map.
 
image

This is the Map I would expect to be shown.

I trust this is the information you required?
What did you want me to do with the <Places-Global.csv> file?

### PR DESCRIPTION
**Pull request by miqrogroove**
Fixes #3229 and improves on the patch from #3291

### CODE REVIEW

**Comment by miqrogroove** [comment]
Just for brainstorming, this could be improved slightly if we have some way to test if the `$features` are not bounded by the saved zoom level.  I haven't tested that so far because it's not an issue on my site.

**Comment by miqrogroove** [comment]
@fisharebest can you confirm for me that the Place->id() value is precached and not generating a query for each call?  I've placed that inside the controller loop and wasn't sure about the DB performance.

**Comment by fisharebest** [comment]
> can you confirm for me that the Place->id() value is precached and not generating a query for each call?

Yes.  See https://github.com/fisharebest/webtrees/blob/2.0.5/app/Place.php#L112-L141

**Comment by miqrogroove** [comment]
That looks like something different.

Let's say I'm calling `->id()` on 300 place objects in one page.  Is this going to run 300 queries, and then cache them?  Or is the id value already cached before I call on it?

**Comment by fisharebest** [comment]
If there are 300 different places, then yes, this will be 300 different calls.

We'll also have 2x300 calls to count the individuals/families at each place.

**Comment by miqrogroove** [comment]
Could you help suggest better code for my line 249?  It's only purpose is to compare place objects.  I'd rather not use the database at all.  😇

File: app/Http/Controllers/PlaceHierarchyController.php [hunk 1]
```diff
@@ -236,16 +236,18 @@ protected function mapData(Place $placeObj): array
...
-            $show_link = false;
-        }
+        // Include parent with its child list so we can see it even if they are in separate locations.
+        $places[] = $placeObj;
 
         foreach ($places as $id => $place) {
             $location = new PlaceLocation($place->gedcomName());
 
+            // Careful, $id != $placeObj->id()
+            $is_child = $place->id() != $placeObj->id();
```
Comment (miqrogroove) [inline app/Http/Controllers/PlaceHierarchyController.php:249]: Needs optimization.  Halp!  😳

Comment (miqrogroove) [reply app/Http/Controllers/PlaceHierarchyController.php]: I think I've figured it out now.  Just using `gedcomName()` instead.

**Comment by miqrogroove** [comment]
This should be all better now.  I didn't realize the ID call was so expensive.  Just using the gedcom string instead.

**Comment by miqrogroove** [comment]
Some parts have been replaced by #3357.  Any remaining bugs will be covered by future patches from the same fork.

**Pull request closed without merging**